ios/PurchasesHybridCommon/PurchasesHybridCommonSwift/CommonFunctionality.swift

```diff
@@ -0,0 +1,512 @@
+    }
+
+    @objc static func setPushToken(_ pushToken: String?) {
+        // todo
+        // this method has been temporarily removed and will be re-added
+        // when this code is adapted for v4
+        // Purchases.shared.setPushToken(pushToken)
+    }
+
+}
+
+// MARK: Attribution IDs
+@objc public extension CommonFunctionality {
+
+    @objc static func collectDeviceIdentifiers() {
+        Purchases.shared.collectDeviceIdentifiers()
+    }
+
+    @objc static func setAdjustID(_ adjustID: String?) {
+        Purchases.shared.setAdjustID(adjustID)
+    }
+    @objc static func setAppsflyerID(_ appsflyerID: String?) {
+        Purchases.shared.setAppsflyerID(appsflyerID)
+    }
+    @objc static func setFBAnonymousID(_ fbAnonymousID: String?) {
+        Purchases.shared.setFBAnonymousID(fbAnonymousID)
+    }
+    @objc static func setMparticleID(_ mParticleID: String?) {
+        Purchases.shared.setMparticleID(mParticleID)
+    }
+    @objc static func setOnesignalID(_ onesignalID: String?) {
+        Purchases.shared.setOnesignalID(onesignalID)
+    }
+    @objc static func setAirshipChannelID(_ airshipChannelID: String?) {
+        Purchases.shared.setAirshipChannelID(airshipChannelID)
+    }
+
+    @objc static func addAttributionData(_ data: [String: Any], network: Int, networkUserId: String) {
+        // todo: clean up force cast after migration to v4
+        Purchases.addAttributionData(data,
+                                     from: RCAttributionNetwork(rawValue: network)!,
+                                     forNetworkUserId: networkUserId)
+    }
+
+}
+
+// MARK: Campaign parameters
+@objc public extension CommonFunctionality {
+
+    @objc static func setMediaSource(_ mediaSource: String?) {
+        Purchases.shared.setMediaSource(mediaSource)
+    }
+    @objc static func setCampaign(_ campaign: String?) {
+        Purchases.shared.setCampaign(campaign)
+    }
+    @objc static func setAdGroup(_ adGroup: String?) {
+        Purchases.shared.setAdGroup(adGroup)
+    }
+    @objc static func setAd(_ ad: String?) {
+        Purchases.shared.setAd(ad)
+    }
+    @objc static func setKeyword(_ keyword: String?) {
+        Purchases.shared.setKeyword(keyword)
+    }
+    @objc static func setCreative(_ creative: String?) {
+        Purchases.shared.setCreative(creative)
+    }
+
+}
+
+private extension CommonFunctionality {
+
+    static func purchaserInfoCompletionBlock(from block: @escaping ([String: Any]?, ErrorContainer?) -> Void)
+    -> ((Purchases.PurchaserInfo?, Error?) -> Void) {
+        return { purchaserInfo, error in
+            if let error = error {
+                let errorContainer = ErrorContainer(error: error, extraPayload: [:])
+                block(nil, errorContainer)
+            } else if let purchaserInfo = purchaserInfo {
+                block(purchaserInfo.dictionary, nil)
+            } else {
+                fatalError("got nil error and nil purchaserInfo")
```

Comment:
should we send an unknown error like in https://github.com/RevenueCat/purchases-hybrid-common/pull/127/files#diff-466f023a265423faf4b0d7c88de6ca5ab6bbb35f1bb0c4b0ec0e1097778fc45cR254

Reply:
yeah, I got inconsistent with unknown error handling there did I. I'll unify, thanks for bringing it up